File: src/2017/ruby/day05.rb
#!/usr/bin/env ruby

def steps(offsets, mut)
    retval = 0
    inc = 0
    ctr = 0
    until ctr >= offsets.size or ctr < 0 do
        retval += 1
        inc = offsets[ctr]
        offsets[ctr] = mut.call(offsets[ctr])
        ctr += inc
    end
    return retval
end

# Read input
$input = ARGF.readlines.map { |s| s.to_i}

# Part 1
puts steps $input.dup, lambda { |x| x + 1 }

# Part 2
puts steps $input.dup, lambda { |x| x < 3 ? x + 1 : x - 1 }

File: src/2017/python/day05.py
#!/usr/bin/env python3

import sys


def steps(offsets, mut):
    retval = 0
    inc = 0
    ctr = 0
    while 0 <= ctr < len(offsets):
        retval += 1
        inc = offsets[ctr]
        offsets[ctr] = mut(offsets[ctr])
        ctr += inc

    return retval


def main():
    offsets = list(map(int, sys.stdin.readlines()))

    # Part 1
    print(steps(list(offsets), lambda x: x + 1))

    # Part 2
    print(steps(list(offsets), lambda x: x + 1 if x < 3 else x - 1))


if __name__ == '__main__':
    sys.exit(main())
